### PR TITLE
Refactor marker connections 2

### PR DESCRIPTION
**Pull request by schaumb**

### CODE REVIEW

**Comment by schaumb** [comment]
There is a crash at the tooltip (markerDC): 
kép
The markerId is not an index of the markers list, but a "unique identifier" for the plot. We need the exact marker position here.

File: src/dataframe/old/types.h [hunk 1]
```diff
@@ -87,11 +92,10 @@ class Filter
 
 	[[nodiscard]] Filter operator&&(const Filter &other) const
 	{
-		auto first =
-		    func1.get() == True ? other.func1.get() : func1.get();
-		auto second =
-		    other.func1.get() == first ? True : other.func1.get();
-		return {first, second};
+		auto &&[min, max] =
+		    std::minmax({func1.get(), other.func1.get()});
```
Comment (schaumb) [inline src/dataframe/old/types.h:96]: [unspecified](https://stackoverflow.com/questions/44397784/is-operator-less-than-on-pointers-consistent) by the compiler. It should use 2nd argument as `std::less<Fun*>` (and [not `less<>`](https://quuxplusone.github.io/blog/2019/01/20/std-less-nightmare/)).

File: src/chart/generator/marker.cpp [hunk 2]
```diff
@@ -122,34 +118,30 @@ Marker::Marker(const Options &options,
...
-
-	if (!first) {
-		double Geom::Point::*const coord =
-		    horizontal ? &Geom::Point::x : &Geom::Point::y;
-		marker.position.*coord += position.*coord;
+	if (prev && next && main && (!first || polarConnection)) {
+		next->prevMainMarker =
+		    MarkerIndexPosition{prev->idx, prev->pos};
+		next->polarConnection = polarConnection && first;
+		return !first || polarConnection;
```
Comment (simzer) [inline src/chart/generator/marker.cpp:131]: always true

File: src/chart/generator/plot.cpp [hunk 3]
```diff
@@ -102,29 +107,109 @@ bool Plot::dimensionMatch(const Plot &a, const Plot &b)
...
+		        smarker.prevMainMarker.values[0].value;
+		    idx != ~Marker::MarkerIndex{}) {
+			if (prePos < source_reindex.size()
+			    && source_reindex[prePos].idx == idx)
+				prePos = source_reindex[prePos].pos;
+			else if (prePos < target_reindex.size()
+			         && target_reindex[prePos].idx == idx)
+				prePos = target_reindex[prePos].pos;
+			else
+				throw std::runtime_error(
```
Comment (simzer) [inline src/chart/generator/plot.cpp:177]: internal error in the message

**Review by simzer — APPROVED** [review]

File: src/chart/rendering/markerrenderer.cpp [hunk 4]
```diff
@@ -419,7 +415,7 @@ std::pair<Gfx::Color, Gfx::Color> MarkerRenderer::getColor(
 
 	const auto &enabled =
 	    label ? abstractMarker.labelEnabled : abstractMarker.enabled;
-	auto alpha = static_cast<double>(enabled) * factor;
+	auto alpha = std::min(static_cast<double>(enabled), factor);
```
[the remark this answers is not in the document]
Comment (schaumb) [reply src/chart/rendering/markerrenderer.cpp:418]: https://github.com/vizzuhq/vizzu-lib/pull/553/files#diff-5b0c755e465d2c5077d9164f0d427678d94f38aa889252d28ba6dbd3ddad49fbR89

File: src/dataframe/old/types.h (hunk 1 above)
Comment (schaumb) [reply src/dataframe/old/types.h:96]: https://github.com/vizzuhq/vizzu-lib/pull/553/files#diff-af29f3d181986e3a0151b57f1598584001d2f9f67668313a4fb2f642e4011deaR96

File: src/chart/generator/marker.cpp (hunk 2 above)
Comment (schaumb) [reply src/chart/generator/marker.cpp:131]: https://github.com/vizzuhq/vizzu-lib/pull/555/files#diff-46a082b5fcd89227b323f2517c3e4cec23ff6d82bb01803c52a3ed33c4d18e8eR131

File: src/chart/generator/plot.cpp (hunk 3 above)
Comment (schaumb) [reply src/chart/generator/plot.cpp:177]: https://github.com/vizzuhq/vizzu-lib/pull/555/files#diff-afcdf1b3efa39a42c582a4c0fed90d990e1a5405760c8c04a1450c2f48eb4536R176-R179

File: src/chart/generator/plotptr.h [hunk 5]
```diff
@@ -22,13 +22,17 @@ struct Buckets
 	{
 		auto &&[i, j] = p;
 		if (k * n < i * j)
-			markers =
-			    std::make_unique_for_overwrite<MarkerPtrArr>(i * j);
+			markers = std::make_unique<MarkerPtrArr>(i * j);
```
[the remark this answers is not in the document]
Comment (schaumb) [reply src/chart/generator/plotptr.h:25]: https://github.com/vizzuhq/vizzu-lib/pull/555/commits/2c3543f6f474e7ba5b3f7686a069887cb70a0e04

This commit is all about it.

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,9 +2,18 @@
 
 ## [Unreleased]
 
+### Fixed
+
+- First marker alpha was different to other marker's alpha
+- Make some transactions clearer. (rectangle-line first marker)
```
[the remark this answers is not in the document]
Comment (schaumb) [reply CHANGELOG.md:8]: https://github.com/vizzuhq/vizzu-lib/pull/555/commits/5486a6f5b60a83dad1327fa4bc85d27fdcc5c850

**Pull request merged**